zrml/court/src/lib.rs

```diff
@@ -104,6 +112,40 @@ mod pallet {
     where
         T: Config,
     {
+        // Returns an unique random subset of `jurors` with length `len`.
+        //
+        // If `len` is greater than the length of `jurors`, then `len` will be capped.
+        pub(crate) fn random_jurors<'a, 'b>(
+            jurors: &'a [(T::AccountId, Juror<BalanceOf<T>>)],
+            len: usize,
+        ) -> ArrayVec<&'b Juror<BalanceOf<T>>, MAX_RANDOM_JURORS>
+        where
+            'a: 'b,
+        {
+            let actual_len = jurors.len().min(len);
+            let mut subset = ArrayVec::new();
+            if actual_len == 0 {
+                return subset;
+            }
+            // https://github.com/paritytech/substrate/issues/8312
+            let (random_hash, _) = T::Random::random(b"zrml-court");
+            for byte in random_hash.as_ref().iter().copied() {
+                if subset.len() == MAX_RANDOM_JURORS {
+                    break;
+                }
+                // `actual_len` will never be 0
+                let idx = Into::<usize>::into(byte) % actual_len;
+                // `idx` will always be within the length of `jurors`
+                let (_, juror) = jurors.get(idx).unwrap();
+                let juror_is_not_included = subset.iter().all(|&el| el != juror);
+                if juror_is_not_included {
+                    // `push` will never overflow the internal capacity of `MAX_RANDOM_JURORS` jurors
+                    subset.push(juror);
+                }
+            }
```

Comment:
This for loop will always select between the first `actual_len` of jurors. If you have for example jurors = [A, B, C, D], and `actual_len = 2`, then it will always select jurors `A` and juror `B`, irrespective of how often and when you will run this function. This is due to the circumstance, that `actual_len` is used for the modulo operation. The following suggestion replaces `actual_len` and adds a break at the end when enough jurors were collected:
```suggestion
            for byte in random_hash.as_ref().iter().copied() {
                if subset.len() == MAX_RANDOM_JURORS {
                    break;
                }
                // `actual_len` will never be 0
                let idx = Into::<usize>::into(byte) % jurors.len();
                // `idx` will always be within the length of `jurors`
                let (_, juror) = jurors.get(idx).unwrap();
                let juror_is_not_included = subset.iter().all(|&el| el != juror);
                if juror_is_not_included {
                    // `push` will never overflow the internal capacity of `MAX_RANDOM_JURORS` jurors
                    subset.push(juror);
                }
                if subset.len() == actual_len { break };
            }
```

The solution above does fix the random selection over the whole set of jurors, but it is not complete at all and must be further improved. Three things have to be considered:
1) If `jurors.len() > 255`, then every juror after index 255 has no chance of being selected (since `a % b = b` for every `b >= a`)
2) The chances are unlikely distributed. Example: We have 120 jurors (this works also with less jurors, but it's easy to see with 120 jurors). If the random byte is between 0 and 119, every juror can be selected. If the random byte is between 120 and 239, every juroro can be selected. If the random byte is between 240 and 255, only the first 15 jurors can be selected! This means, the first fifteen jurors have a higher likelyhood to be selected as jurors and will be present in more courts than the other jurors in the long term.
3) If there are more than `byte_count(random_hash)` jurors to be selected, then this function will cap out at max `byte_count(random_hash)` jurors. For example, if we want 260 jurors, max 256 jurors will be selected and 4 will be left out. Even worse: If we want 120 jurors for example, and the loop fails to find 120 distinguished jurors after iterating over every random byte, the subset will contain less than 120 jurors. A solution to that would be to generate new randomness and start another loop to find the remaining jurors.

Reply:
One simple approach towards ensuring almost the same likelyhood for every juror to be selected, is to use the whole random number and execute modulo on it. The likelyhood will still be uneven, but the difference is so incredibly small, that practically this effect will not be observable.